hydromt/components/table.py

```diff
@@ -0,0 +1,126 @@
+"""Table component."""
+
+import glob
+import os
+from os.path import basename, dirname, join
+from typing import (
+    TYPE_CHECKING,
+    Dict,
+    Optional,
+    Union,
+    cast,
+)
+
+import pandas as pd
+
+from hydromt.components.base import ModelComponent
+from hydromt.hydromt_step import hydromt_step
+
+if TYPE_CHECKING:
+    from hydromt.models.model import Model
+
+__all__ = ["TableComponent"]
+
+
+class TableComponent(ModelComponent):
+    """Table Component."""
+
+    def __init__(
+        self,
+        model: "Model",
+    ):
+        """Initialize a TableComponent.
+
+        Parameters
+        ----------
+        model: Model
+            HydroMT model instance
+        """
+        self._data: Optional[Dict[str, Union[pd.DataFrame, pd.Series]]] = None
+        super().__init__(model=model)
+
+    @property
+    def data(self) -> Dict[str, Union[pd.DataFrame, pd.Series]]:
+        """Model tables."""
+        if self._data is None:
+            self._initialize_tables()
+        if self._data is None:
+            raise RuntimeError("Could not load data for table component")
+        else:
+            return self._data
+
+    def _initialize_tables(self, skip_read=False) -> None:
+        """Initialize the model tables."""
+        if self._data is None:
+            self._data = dict()
+            if self._root.is_reading_mode() and not skip_read:
+                self.read()
+
+    @hydromt_step
+    def write(self, fn: str = "tables/{name}.csv", **kwargs) -> None:
+        """Write tables at <root>/tables."""
+        self._root._assert_write_mode()
+        if self.data:
```

Comment:
`self.data` will always initialize to an empty dict. So it will always write. Not sure if this is the correct check. Perhaps check on `self._data`

Reply:
good point

Reply:
So there's a reason why we check on ``self.data`` and not ``self._data`` here.
In case of update, ``self.tables`` might never be updated so may never be initialized. When a user specifically wants to write, then it should write without the user having to manually call ``self.tables.read`` or just ``self.tables`` first. This led to a lot of unexpected bugs in the past.
What we use to do but was maybe not well updated in core in Model.write is that in that function we check on _tables to only write components that actually were initialized. See the hydromt_wflow example:

```
    def write(self):
        """Write the complete model schematization and configuration to file."""
        self.logger.info(f"Write model data to {self.root}")
        # if in r, r+ mode, only write updated components
        if not self._write:
            self.logger.warning("Cannot write in read-only mode")
            return
        self.write_data_catalog()
        if self.config:  # try to read default if not yet set
            self.write_config()
        if self._grid:
            self.write_grid()
        if self._tables:
            self.write_tables()
        if self._geoms:
            self.write_geoms()
        if self._forcing:
            self.write_forcing()

```

So please revert back to a check on ``self.data`` but maybe a better check that the dict is not empty.